Gmail API journal submit system

daisieh opened the pull request:
First draft of code to call Gmail's API directly to retrieve journal submission emails from the dryad.journal.submit@gmail.com email account.

In order to enable this functionality, you need to insert the json clientsecret in your machine's maven settings.xml. An example can be found at dev.datadryad.org:/home/dhdryad/gmail.settings.xml.

Testing and detailed configuration instructions can be found on the wiki: http://wiki.datadryad.org/Journal_Metadata_Processing_Technology

dleehr commented:
#### Code Review Checklist
- [x] Read through the code to ensure it is:
- [x] Legible
- [x] Well-commented - **mostly yes**, a couple of misleading comments to cleanup
- [ ] Free of any "code smells" - recommended a couple changes
- [ ] Ensure the associated documentation is up to date. **needs minor clarifications**
- [x] Compile it on your local machine and test the new functionality. **core functionality is working, some feedback included**
- [ ] Verify that the change is accompanied by adequate testing code. Verify that the code passes the tests. - there are no automated tests

dleehr commented on dspace/config/dspace.cfg:
```diff
@@ -2478,6 +2478,13 @@ solr.default.filter.override=location
+
+# Copy the json line for this property from dev.datadryad.org/gmail.settings.xml to your maven settings.xml
+submit.journal.clientsecrets = ${default.submit.journal.clientsecret}
+submit.journal.email = dryad.journal.submit@gmail.com
```
This should be pulled from the maven profile rather than hard-coded

dleehr commented on dspace/config/dspace.cfg:
```diff
@@ -2478,6 +2478,13 @@ solr.default.filter.override=location
+# Copy the json line for this property from dev.datadryad.org/gmail.settings.xml to your maven settings.xml
+submit.journal.clientsecrets = ${default.submit.journal.clientsecret}
+submit.journal.email = dryad.journal.submit@gmail.com
+submit.journal.email.label = journal-submit
```
Also should be pulled from maven profile

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -25,22 +25,14 @@
 
+import java.io.*;
+
+import java.lang.Exception;
```
There's no need to import anything from java.lang

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -70,162 +68,98 @@
+        } catch (Exception details) {
             sendEmailIfConfigured(details);
 
             if (details instanceof SubmissionException) {
```
I know this is pre-existing code, but this is clunky exception handling

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -271,25 +205,20 @@ public void init(ServletConfig aConfig) throws ServletException {
+                ConfigurationManager.loadConfig(config);
             }
 
             String journalPropFile = ConfigurationManager.getProperty("submit.journal.config");
```
Just a heads up, but this will need to be updated to mesh with #721

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -397,96 +323,186 @@ private void sendEmailIfConfigured(Exception aException)
+            throw new SubmissionException(result.getStatus());
+        }
+
+        // isHas?
```
I can isHas better method name (I know it's been here a while)

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -535,14 +548,46 @@ public String getServletInfo() {
+        @Override
+        public void run() {
+            ArrayList<String> labels = new ArrayList<String>();
+            labels.add("journal-submit");
```
This should be read from the properties file since it is defined there.

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -535,14 +548,46 @@ public String getServletInfo() {
+                    LOGGER.info("got " + messages.size() + " messages");
+                    for (Message message : messages) {
+                        LOGGER.info("Message: "+ message.getId() + ", " + message.getSnippet());
+                        ByteArrayInputStream postBody = new java.io.ByteArrayInputStream(Base64.decodeBase64(message.getRaw()));
```
Since you've already encapsulated the Gmail communication in DryadGmailService, it may be a good idea to encapsulate the messages in a class too - and handle the base64 decoding and other Gmail implementation-specific details there.

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadGmailService.java:
```diff
@@ -0,0 +1,213 @@
+        }
+    }
+
+    // Default constructor builds a DryadGmailService for the dryad-journal-submit@gmail.com account
```
well it builds a service for the account named by the properties file

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadGmailService.java:
```diff
@@ -0,0 +1,213 @@
+        DryadGmailService dryadGmailService = new DryadGmailService();
+        // TEST CASE:
+        ArrayList<String> labels = new ArrayList<String>();
+        labels.add("test_gmail_api");
```
Magic string

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -70,162 +68,98 @@
+            try {
+                dryadGmailService = new DryadGmailService();
+                ArrayList<String> labels = new ArrayList<String>();
+                labels.add("test_gmail_api");
```
magic string

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -351,30 +274,33 @@ else if (property.equals("fullname")) {
+        LOGGER.debug("scheduling email harvesting");
+        myEmailHarvester = new Timer();
+        // schedule email harvesting to happen once a day
+        myEmailHarvester.schedule(new DryadEmailSubmissionHarvester(), 0, 1000 * 60 * 60 * 24);
```
Once a day is probably not frequent enough. The submitter will have the link to submit their data as soon as we do, so we probably don't want to introduce a delay of up to 24h. This should be configurable.
Also, having an on-demand option would be good too, . The current process is instantaneous, and when working through setting up new journals, I think that tight feedback loop helps.

dleehr commented:
Some broader thoughts:

Currently the script that POSTs incoming emails will deliver it to both dev and prod. So both servers get copies of the XML. This allows us to work with journals and tweak their email formats while checking on dev or staging, before they're enabled on prod.

Since this implementation looks for the `journal-submit` label and then takes it off, I don't see how dev and prod could both run this simultaneously. One way around that would be to use a different label for dev, ensure each server only uses its own label (see earlier comments), and assign all the labels on receipt. Or skip the pre-assignment and each server marks the message with its own done label.

One note on the wiki documentation - the testing section still mentions POSTing with cURL - should clarify this to indicate what's old and what's current

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadGmailService.java:
```diff
@@ -0,0 +1,213 @@
+        return result;
+    }
+
+    public List<Message> retrieveMessagesWithLabels (List<String> labels) throws IOException {
```
If the labels don't actually exist, this will return all messages. I discovered this by trying to GET `/journal-submit/test` without having a `test_gmail_api` label defined. It looked like it was running forever, but eventually logged everything in my Gmail account

dleehr commented on dspace/modules/journal-submit/src/main/java/org/datadryad/submission/DryadEmailSubmission.java:
```diff
@@ -70,162 +68,98 @@
+                dryadGmailService = new DryadGmailService();
+                ArrayList<String> labels = new ArrayList<String>();
+                labels.add("test_gmail_api");
+                List<Message> messages = dryadGmailService.retrieveMessagesWithLabels(labels);
```
When the labels here don't exist in the gmail account, this method prints out every message to the log and does not send a response

dleehr commented:
The harvester doesn't run until the servlet is initialized, which only happens on the first request. If you want to initialize it when tomcat starts, add a load-on-startup tag to [web.xml](https://github.com/datadryad/dryad-repo/blob/bd4d18d47409ba068ddffc2df65d817077ab7714/dspace/modules/journal-submit/src/main/webapp/WEB-INF/web.xml#L27):

```
<servlet>
  ...
  <load-on-startup>1</load-on-startup>
</servlet>
```

Otherwise we won't process any emails until something makes a request to the servlet.

dleehr commented:
Back to you, @daisieh.

dleehr commented on dspace/config/dspace.cfg:
```diff
@@ -2478,6 +2478,13 @@ solr.default.filter.override=location
+submit.journal.credential.dir = ${dspace.dir}/submission/credential
+
+# Copy the json line for this property from dev.datadryad.org/gmail.settings.xml to your maven settings.xml
+submit.journal.clientsecrets = ${default.submit.journal.clientsecret}
```
Can you explain what this is and how to generate it? Presumably it's linked to someone's google account and could be revoked.

daisieh commented:
I think I addressed all of the comments here *except* for the one about how to deal with both dev and prod needing to be able to tag/untag messages manually.

While thinking about this, I also realized that it is probably not best, when testing, to actually process/untag journal-submit messages in the gmail account unless we really did process them on prod. Therefore, I pushed the label out to the maven settings so that you have to manually set the label that will be processed and unlabeled. So if you wanted to test an integration with a journal on dev, you could label some old test emails with the dev server's designated label, like "dev-test-label" or something (through the Gmail website) and then process them and watch the emails get tagged and untagged.

So, the only maven settings.xml that should have the label "journal-submit" is prod: `        <default.submit.journal.label>journal-submit</default.submit.journal.label>` Other servers should have other labels.

daisieh commented:
In re: old workflow still on wiki, like the POST stuff: I will remove that when we push this change out to production.

daisieh commented:
I think I addressed all of the comments here _except_ for the one about how to deal with both dev and prod needing to be able to tag/untag messages manually.

While thinking about this, I also realized that it is probably not best, when testing, to actually process/untag journal-submit messages in the gmail account unless we really did process them on prod. Therefore, I pushed the label out to the maven settings so that you have to manually set the label that will be processed and unlabeled. So if you wanted to test an integration with a journal on dev, you could label some old test emails with the dev server's designated label, like "dev-test-label" or something (through the Gmail website) and then process them and watch the emails get tagged and untagged.

So, the only maven settings.xml that should have the label "journal-submit" is prod: `<default.submit.journal.label>journal-submit</default.submit.journal.label>` Other servers should have other labels.

dleehr commented:
> Therefore, I pushed the label out to the maven settings so that you have to manually set the label that will be processed and unlabeled. So if you wanted to test an integration with a journal on dev, you could label some old test emails with the dev server's designated label, like "dev-test-label" or something (through the Gmail website) and then process them and watch the emails get tagged and untagged.

Thanks. I think that will help with the two servers fighting each other - we definitely don't want to miss emails on prod because dev processed them, so that's good. I'd also try to make this work as close to the previous behavior as possible - we should probably add the dev-test-label (or whatever it's called) automatically to every message and maybe even shorten the timer duration further (on dev anyways)

dleehr commented:
> So if you wanted to test an integration with a journal on dev, you could label some **old** test emails with the dev server's designated label

Just a clarification on this - there wouldn't be **old** test emails for a new journal we were bringing online. The journal would be sending test emails and someone (@jamie-dryad) would be seeing the new messages coming in and checking on the dev server for a valid XML file. If the XML is missing anything or the message doesn't get parsed, she'll work it out with them and they'll send another.

So this is one way the current system works well - the XML file gets created immediately, and there's no extra step to see it on dev.

dleehr commented:
Some minor documentation updates needed, including how to use this on dev. Otherwise, good to merge

daisieh commented:
@dleehr: Do you mean wiki documentation or doc within the code?

dleehr commented:
Wiki
